docs/source/hooks.rst

```diff
@@ -0,0 +1,173 @@
+Hooks make it possible to track state changes to each job in a **signac** project as the :ref:`FlowProject<flow-project>` operates on it.
+
+A hook is a function that is called at a specific point relative to the execution of a **signac-flow** :ref:`operation<operations>`.
+A hook is triggered when an operation starts, exits, succeeds, or raises an exception.
+
+A basic use case is to log the success/failure of an operation by creating a hook that sets a job document value ``job.doc.operation_success`` to ``True`` or ``False``.
+Additionally, a user may record the `git commit ID <https://git-scm.com/book/en/v2/Git-Basics-Viewing-the-Commit-History>`_ upon the start of an operation, allowing them to track which version of code ran the operation.
+
+.. _hook_triggers:
+
+Triggers
+=============
+
+The following triggers are provided:
+
+1. :py:meth:`~flow.FlowProject.operation_hooks.on_start` will execute when the operation begins execution.
+2. :py:meth:`~flow.FlowProject.operation_hooks.on_exit` will execute when the operation exits, with or without error.
+3. :py:meth:`~flow.FlowProject.operation_hooks.on_success` will execute when the operation exits without error.
+4. :py:meth:`~flow.FlowProject.operation_hooks.on_exception` will execute when the operation exits with error.
+
+Hooks can be installed at the :ref:`operation level <operation-hooks>` or at the :ref:`flow-project level<project-level-hooks>`.
+Project-level hooks are called for every operation in the flow project.
+
+The hooks created with triggers :py:meth:`~flow.FlowProject.operation_hooks.on_start`, :py:meth:`~flow.FlowProject.operation_hooks.on_exit`,  and :py:meth:`~flow.FlowProject.operation_hooks.on_success` require two arguments: the operation name and the :py:class:`signac.contrib.job.Job` object. Hooks created to trigger :py:meth:`~flow.FlowProject.operation_hooks.on_exception` require three arguments: the operation name, the output error, and the job object.
+
+.. note::
+
+    Hooks are run in the environment of the python process from which you call FlowProject.main().
+    For this reason, hooks will not have access to modules in a container specified in the :term:`executable directive<executable>`.
+
+.. _operation-hooks:
+
+Operation Hooks
+===============
+
+Hooks may be added to individual operations using decorators.
+The :py:class:`~flow.FlowProject.operation_hooks` decorator tells :py:class:`~signac` to run a hook (or set of hooks) when an operation reaches the specified trigger.
+
+The :py:class:`~flow.FlowProject.operation_hooks` decorator accepts objects as a function of the job operation (:py:class:`~flow.project.JobOperation`).
+
+
+An operation hook can be used to store basic information about the execution of a job operation in the job document.
+In the following example, when our test operation ``error_on_a_0`` generates an error, the hook function ``store_error_to_doc`` executes.
```

Comment:
This document bounces around a fair amount between using plural first and second-person we/you/our and avoiding them. I think it might be best to avoid these words based on the informational style of this document.

Also I prefer "raise an exception" instead of "generate an error."
```suggestion
In the following example, when the test operation ``error_on_a_0`` raises an exception, the hook function ``store_error_to_doc`` executes.
```